### PR TITLE
Rerun Apply Shim when mixins with consumers are redefined

### LINKED ISSUE #3802: ApplyShim can produce incorrect properties when elements are shared
See http://jsbin.com/qipodaq/edit?html,output. Inside `x-host2`, `my-element` should have a green background. 

In this case `x-mixin` contains a rule `@apply --mixin`. It is used within `x-host1` and `x-host2` but each of these defines a different set of properties for `--mixin`. Whichever of these elements stamps last will have incorrect properties applied via `@apply --mixin`. This is because we "bake" the stylesheet for `x-mixin` at the time its first instance is created.

Here's a proposal for how to fix this:

- For every mixin name, the ApplyShim keeps a map of elements that depend on (`@apply`) that mixin.
- Whenever the set of properties for a mixin is added to, each of the elements in the elements map is invalidated.
- Whenever an element is created, if it has already run the ApplyShim, it checks to see if any of its mixin dependencies are invalid and if so, it re-runs the ApplyShim. For ShadyDOM, this can replace the element's stylesheet. For ShadowDOM, it is probably good enough to fix the stylesheet in the element's template thus affecting all subsequent instances.

Sketch:

When `x-host1` is defined:

```
MixinMap:
  '--mixin': {
    properties: {color: true},
    elements: {'x-mixin': true}
  }
```

When `x-host2` is defined:

Add `background: green` to `--mixin` and invalidate `x-mixin`
 
```
MixinMap:
  '--mixin': {
    properties: {color: true, background: true},
    elements: {'x-mixin': false}
  }
```

When the next `x-mixin` is created:

- IF it has already run the ApplyShim, 
- then get the list of mixins it uses [`--mixin`], 
- check if any of these are invalid `(MixinMap['--mixin'].elements['x-mixin'] === false)`
 - if yes, re-run the ApplyShim

### PR DESCRIPTION
**Pull request by azakus**
When element stamps, make its mixins are valid, and if not, run the
Apply Shim again.

Flag is stored on element prototype to make the "valid" check extremely
fast

Only new instances of consuming elements will be updated

Fixes #3802

### CODE REVIEW

File: polymer.html [hunk 1]
```diff
@@ -72,6 +72,7 @@
       this._setupShady();
       this._registerHost();
       if (this._template) {
+        this._validateMixins();
```
Comment (sorvell) [inline polymer.html]: Let's name this more explicitly, what about '_validateApplyShim'

File: src/lib/apply-shim.html [hunk 2]
```diff
@@ -133,12 +136,17 @@
...
     var mixinValues = cssTextToMap(mixinAsProperties);
-    var oldProperties = mapGet(propertyName);
     var combinedProps = mixinValues;
-    if (oldProperties) {
+    var mixinEntry = mapGet(propertyName);
+    if (mixinEntry) {
       // NOTE: since we use mixin, the map of properties is updated here
       // and this is what we want.
-      combinedProps = Polymer.Base.mixin(oldProperties, mixinValues);
+      combinedProps = Polymer.Base.mixin(mixinEntry.properties, mixinValues);
```
Comment (sorvell) [inline src/lib/apply-shim.html]: We only have to invalidate if `mixinValues` adds new keys. Do you think it's worth detecting that?

Comment (azakus) [inline src/lib/apply-shim.html]: I think that can be done in the loop lower that generates `initial` values. I'll give it a shot.

File: src/lib/apply-shim.html [hunk 3]
```diff
@@ -133,12 +136,17 @@
...
+      for (var elementName in mixinEntry.dependants) {
+        if (elementName !== ApplyShim.__currentElementProto.is) {
```
Comment (sorvell) [inline src/lib/apply-shim.html]: Is it too inconvenient to thread through the current element name here?

File: src/lib/apply-shim.html [hunk 4]
```diff
@@ -118,6 +121,16 @@
     return out;
   }
 
+  function invalidateMixinEntry(mixinEntry) {
+    var currentProto = ApplyShim.__currentElementProto;
+    var currentElementName = currentProto && currentProto.is;
+    for (var elementName in mixinEntry.dependants) {
+      if (elementName !== currentElementName) {
+        mixinEntry.dependants[elementName].__mixinsInvalid = true;
```
Comment (sorvell) [inline src/lib/apply-shim.html]: Let's also make this flag more explicit, maybe `__applyShimInvalid`

File: src/lib/custom-style.html [hunk 5]
```diff
@@ -209,6 +209,7 @@
       }
       var styleRules = styleUtil.rulesForStyle(e);
       if (!targetedBuild) {
+        applyShim.__currentElementProto = this.__proto__;
```
Comment (sorvell) [inline src/lib/custom-style.html]: What use case does this support?

**Comment by sorvell** [comment]
LGTM

**Pull request merged**